identity/graphql.go

```diff
@@ -0,0 +1,276 @@
+	case args.Signature != nil:
+		rows, err = r.db.Query("SELECT * FROM claim WHERE signature = ?", *args.Signature)
+	case args.ID != nil:
+		rows, err = r.db.Query("SELECT * FROM claim WHERE id = ?", *args.ID)
+	default:
+		return nil, errors.New("missing issuer,subject,claim,signature or id argument")
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var resolvers []*ClaimResolver
+	for rows.Next() {
+		var claim Claim
+		if err := rows.Scan(&claim.Issuer, &claim.Subject, &claim.Claim, &claim.Signature, &claim.ID); err != nil {
+			return nil, err
+		}
+		resolvers = append(resolvers, &ClaimResolver{&claim})
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	return resolvers, nil
+}
+
+// ClaimResolver defines GraphQL resolver functions for Claim fields.
+type ClaimResolver struct {
+	claim *Claim
+}
+
+// Issuer resolver
+func (c *ClaimResolver) Issuer() string {
+	return c.claim.Issuer
+}
+
+// Subject resolver
+func (c *ClaimResolver) Subject() string {
+	return c.claim.Subject
+}
+
+// Claim resolver
+func (c *ClaimResolver) Claim() string {
+	return c.claim.Claim
+}
+
+// Signature resolver
+func (c *ClaimResolver) Signature() string {
+	return c.claim.Signature
+}
+
+// ID resolver
+func (c *ClaimResolver) ID() string {
+	return c.claim.ID
+}
+
+// CreateIdentityArgs are the arguments for a GraphQL CreateIdentity mutation.
+type CreateIdentityArgs struct {
+	Username  *string
+	Owner     *string
+	Signature *string
+}
+
+// CreateIdentity is a GraphQL resolver function which create ID object, store and
+// index it on SQLite3 index.
+func (r *Resolver) CreateIdentity(args *CreateIdentityArgs) (*IdentityResolver, error) {
+	if args.Owner == nil || args.Username == nil || args.Signature == nil {
+		return nil, errors.New("CreateIdentity: one or more argument is nil")
+	}
+	metaid, err := NewIdentity(*args.Username, common.HexToAddress(*args.Owner), common.FromHex(*args.Signature))
+	if err != nil {
+		return nil, err
+	}
+
+	index, err := r.store.OpenIndex("id.index.meta")
+	if err != nil {
+		return nil, err
+	}
+	indexer, err := NewIndexer(index)
+	if err != nil {
+		return nil, err
+	}
+	if err := indexer.IndexIdentity(metaid); err != nil {
+		return nil, err
+	}
+	return &IdentityResolver{identity: metaid}, nil
+}
+
+// CreateClaimArgs are the arguments for a GraphQL CreateClaim mutation.
+type CreateClaimArgs struct {
+	Issuer    *string
+	Subject   *string
+	Claim     *string
+	Signature *string
+}
+
+// CreateClaim is a GraphQL resolver function which create claim object, store and
+// index it on SQLite3 index.
+func (r *Resolver) CreateClaim(args *CreateClaimArgs) (*ClaimResolver, error) {
+
+	if args.Issuer == nil || args.Subject == nil || args.Claim == nil || args.Signature == nil {
+		return nil, errors.New("CreateClaim: one or more argument is nil")
+	}
+	claim := NewClaim(*args.Issuer, *args.Subject, *args.Claim, *args.Signature)
+	index, err := r.store.OpenIndex("claim.index.meta")
```

Comment:
Shouldn't this be the same index as the `CreateIdentity` call? I think the `meta.Index` should be passed to the resolver (rather than the resolver opening an index every time).

Reply:
updated at 19ebf27